src/deluge/gui/views/audio_clip_view.cpp

```diff
@@ -324,7 +325,46 @@ ActionResult AudioClipView::buttonAction(deluge::hid::Button b, bool on, bool in
 	}
 
 	else if (b == X_ENC) {
-		goto dontDeactivateMarker;
+		// removing time stretching by re-calculating clip length based on length of audio sample
+		if (Buttons::isShiftButtonPressed()) {
+			if (on && currentUIMode == UI_MODE_NONE) {
+				if (inCardRoutine) {
+					return ActionResult::REMIND_ME_OUTSIDE_CARD_ROUTINE;
+				}
+
+				AudioClip* audioClip = getCurrentAudioClip();
+				SamplePlaybackGuide guide = audioClip->guide;
+				SampleHolder* sampleHolder = (SampleHolder*)guide.audioFileHolder;
+				if (sampleHolder) {
+					Sample* sample = (Sample*)sampleHolder->audioFile;
+					if (sample) {
+						// get sample start position (green waveform marker)
+						int32_t loopStartSample =
```

Comment:
can this just use SampleHolder::getLengthInSamplesAtSystemSampleRate()? This is recovering it from the actual sample file, and I believe it will be wrong if the sample is not 16bit/44khz